### DISCUSSION TITLE
Issues connecting to remote server

### DISCUSSION THREAD

**Discussion by johnaohara** [pgmoneta/pgmoneta#182] [Q&A] [answered]
Hey @jesperpedersen hope you are doing well!!

I have made a couple of changes to our setup and ran into some problems, I hope you can point me in the correct direction!

The main change has been to try running pgmoneta to backup a remote postgres instance, i.e. not an instance on localhost

After changing the config to change the hostname (and making sure the repl user could connect from the remote machine etc)  I was seeing the following message in the pgmoneta logs;

`Server primary need wal_level at replica or logical`

To bring our install up to date, I rebuilt pgmoneta from the current main branch and restarted, but now I am seeing when I try to start;

`2024-01-29 05:27:48 FATAL main.c:1882 Could not get version for server primary`

I have checked the version query : (https://github.com/pgmoneta/pgmoneta/blob/main/src/libpgmoneta/server.c#L271)  using a remote log in for the repl user and it is working as expected;

```
postgres=> SELECT split_part(split_part(version(), ' ', 2), '.', 1);
 split_part 
------------
 16
(1 row)
```

So questions...

1 - There is no info about why the server has failed validation, so idk what i need to change/look for to rectify. Any ideas?
2 - any pointers as to why the version check query failed? could this be a bug?

Many Thanks :)

**Comment by Jubilee101** [comment]
Hi @johnaohara ,

For 1, perhaps you should try checking the wal_level in postgresql.conf, please see details [here](https://github.com/pgmoneta/pgmoneta/blob/main/doc/tutorial/01_install.md#make-sure-that-replication-level-is-set). 

For 2, it is a little bizarre. I think you are connecting to the server just fine, otherwise it won't start querying the server in the first place. For `pgmoneta_server_get_version()` to return error code 1, most likely something is wrong with pgmoneta_query_execute(), and/or you are getting NULL query_response. Could you check in which step inside [pgmoneta_query_execute()](https://github.com/pgmoneta/pgmoneta/blob/0bbebedd05d968dd6143c687067e1649d1cd55dd/src/libpgmoneta/message.c#L1030) pgmoneta entered `error` section?

**Reply by johnaohara** [reply]
HI @Jubilee101 thanks for your response.

1)  The instance that I am trying to connect to has been created from a backup that has been created from pgmoneta, just on a different machine. The configuration of the server is the same as it was when running as a local instance;

```
$ grep wal_level  /db/path/primary/postgresql.conf 
wal_level = replica			# minimal, replica, or logical
```

2) I will dig a bit more to see where in that function the failure occurs

**Reply by johnaohara** [reply]
@Jubilee101 it is this line that is returning an error: https://github.com/pgmoneta/pgmoneta/blob/0bbebedd05d968dd6143c687067e1649d1cd55dd/src/libpgmoneta/message.c#L1086

**Reply by johnaohara** [reply]
I need to capture the packets to see what is happening on the wire, obviously the database is not returning the expected response

**Reply by Jubilee101** [reply]
Hmm that means the server is not sending us row description messages. It would be nice if you could check message type it's sending us:
https://github.com/pgmoneta/pgmoneta/blob/0bbebedd05d968dd6143c687067e1649d1cd55dd/src/libpgmoneta/utils.c#L310

I suspect it's an error response. We've got functions to parse errors.

**Reply by Jubilee101** [reply]
On second thought it's unlikely to be an error response, we would've caught it here:
https://github.com/pgmoneta/pgmoneta/blob/0bbebedd05d968dd6143c687067e1649d1cd55dd/src/libpgmoneta/message.c#L1081

**Reply by johnaohara** [reply]
It is returning warning messages, before returning the result. I am attempting to capture the packets on the remote machine with TCP dump to see what postgres is sending across the wire, but it's been a while since I  have done that remotely

**Reply by Jubilee101** [reply]
Thank you so much. I'll wait for the update. Is there anything on postgres' log that's worth noticing?

**Reply by johnaohara** [reply, marked as answer]
This is the best I can do atm: https://gist.github.com/johnaohara/ea65c0b13fab28f0e05f1b5d5eff04e9

On the server, I see the same warnings that is sent accross the wire: 

```
25985 [2024-01-31 01:13:17.540 EST] [] [0] WARNING:  database "postgres" has a collation version mismatch
25985 [2024-01-31 01:13:17.540 EST] [] [0] DETAIL:  The database was created using collation version 2.28, but the operating system provides version 2.36.
25985 [2024-01-31 01:13:17.540 EST] [] [0] HINT:  Rebuild all objects in this database that use the default collation and run ALTER DATABASE postgres REFRESH COLLATION VERSION, or build PostgreSQL with the right library version.
```

I will try updating the collation on the remote machine

**Reply by Jubilee101** [reply]
Hmm, seems that this error response message is not picked up by pgmoneta. It should've gone to `error` before `pgmoneta_extract_message_from_data`. And we should definitely parse the message into log so that you wouldn't have to gone through any of this.

Anyway, let me know if that solves the problem. I'll do some further checking and add more informative logs.

**Comment by johnaohara** [comment]
> It should've gone to error before pgmoneta_extract_message_from_data

The server is actually returning the response on the wire.  maybe the response buffer was not big enough, given the warning messages that were returned? in a normal scenario, you would not have the warnings returned

> And we should definitely parse the message into log so that you wouldn't have to gone through any of this.

This would help a lot to track down the issue

**Reply by johnaohara** [reply]
It looks like there are multiple messages returned, and pgmoneta is only inspecting the first message returned on the socket, but the actual **correct** result is in a later message

**Reply by Jubilee101** [reply]
> The server is actually returning the response on the wire. maybe the response buffer was not big enough, given the warning messages that were returned? in a normal scenario, you would not have the warnings returned

Yes, that's a good point. I'll talk with @jesperpedersen to see how to fix this. We have new stream buffer that's more adaptive. And we should definitely skip warning messages.

> This would help a lot to track down the issue

Will have that in the near future!

> It looks like there are multiple messages returned, and pgmoneta is only inspecting the first message returned on the socket, but the actual correct result is in a later message

Yeah, probably need some good fixes in the query execution functionality. We'll submit a patch as soon as possible. 

Thank you so much, and so sorry for the trouble. I assume updating the collation fixed the problem?

**Reply by johnaohara** [reply]
Thanks a lot for your help!
